src/projection_client.rs

```diff
@@ -0,0 +1,629 @@
+use crate::event_store::client::projections;
+use crate::event_store::client::shared::Empty;
+use crate::grpc::{ClientSettings, GrpcClient};
+use crate::Credentials;
+use futures::stream::BoxStream;
+use serde::de::DeserializeOwned;
+
+#[derive(Clone, Debug)]
+pub enum StatsFor {
+    Name(String),
+    AllProjections,
+    AllTransients,
+    AllContinuous,
+    AllOneTimes,
+}
+
+#[derive(Clone, Debug)]
+pub struct StatsDetails {
+    pub core_processing_time: i64,
+    pub version: i64,
+    pub epoch: i64,
+    pub effective_name: String,
+    pub writes_in_progress: i32,
+    pub reads_in_progress: i32,
+    pub partitions_cached: i32,
+    pub status: String,
+    pub state_reason: String,
+    pub name: String,
+    pub mode: String,
+    pub position: String,
+    pub progress: f32,
+    pub last_checkpoint: String,
+    pub events_processed_after_restart: i64,
+    pub checkpoint_status: String,
+    pub buffered_events: i64,
+    pub write_pending_events_before_checkpoint: i32,
+    pub write_pending_events_after_checkpoint: i32,
+}
+
+#[derive(Clone)]
+pub struct ProjectionClient {
+    client: GrpcClient,
+    settings: ClientSettings,
+}
+
+impl ProjectionClient {
+    pub async fn create(settings: ClientSettings) -> Result<Self, Box<dyn std::error::Error>> {
+        let client = GrpcClient::create(settings.clone()).await?;
+
+        Ok(ProjectionClient { client, settings })
+    }
+
+    pub async fn create_one_time_projection(
+        &self,
+        query: String,
+        credentials: Option<&Credentials>,
+    ) -> crate::Result<()> {
+        self.create_projection(
+            credentials,
+            projections::create_req::Options {
+                query,
+                mode: Some(projections::create_req::options::Mode::OneTime(Empty {})),
+            },
+        )
+        .await
+    }
+
+    pub async fn create_transient_projection<Name>(
+        &self,
+        name: Name,
+        query: String,
+        credentials: Option<&Credentials>,
+    ) -> crate::Result<()>
+    where
+        Name: AsRef<str>,
+    {
+        self.create_projection(
+            credentials,
+            projections::create_req::Options {
+                query,
+                mode: Some(projections::create_req::options::Mode::Transient(
+                    projections::create_req::options::Transient {
+                        name: name.as_ref().to_string(),
+                    },
+                )),
+            },
+        )
+        .await
+    }
+
+    pub async fn create_continuous_projection<Name>(
+        &self,
+        name: Name,
+        query: String,
+        track_emitted_streams: bool,
```

Comment:
**CONCERN:** In .NET client `track_emitted_streams` is optional and defaulted to `false`.

Reply:
Rust doesn't not support optional parameters. The reason the update version uses Option is because we don't want to override the previous value.

Reply:
Shouldn't we use options classes as we did in other places for optional parameters?

Reply:
Done